vm/src/obj/objsuper.rs

```diff
@@ -53,27 +54,30 @@ impl PySuper {
     #[pymethod(name = "__getattribute__")]
     fn getattribute(&self, name: PyStringRef, vm: &VirtualMachine) -> PyResult {
         let inst = self.obj.clone();
-        let typ = self.typ.clone();
-
-        match typ.payload::<PyClass>() {
-            Some(PyClass { ref mro, .. }) => {
-                for class in mro {
-                    if let Ok(item) = vm.get_attribute(class.as_object().clone(), name.clone()) {
-                        if item.payload_is::<PyBoundMethod>() {
-                            // This is a classmethod
-                            return Ok(item);
-                        }
-                        return vm.call_if_get_descriptor(item, inst.clone());
-                    }
+
+        // Super should be the next matching class in the object original class' mro after the current one.
+        let mut mro_iter = match self.obj_type.payload::<PyClass>() {
+            Some(PyClass { ref mro, .. }) => mro.iter(),
+            _ => panic!("not Class"),
+        };
+        // The type itself is not in its mro, so skip finding the current class if its the type.
+        if !self.typ.is(&self.obj_type) {
```

Comment:
Can we move this validation to the creation of the super?

Reply:
This is more than validation, it also shifts the mro iterator to the correct location ahead. It should probably also be checked in the constructor but it is part of the code here as well.

Reply:
I have decided to simply do the checks at the constructor, calculate everything there, and remember the resulting mro.